typing/typecore.ml

```diff
@@ -2005,6 +2025,8 @@ and type_pat_aux
         pat_env = !!penv;
         pat_attributes = sp.ppat_attributes;
       }
+  | Ppat_effect _ ->
+      raise (Error (loc, !!penv, Effect_pattern_below_toplevel))
```

Comment:
This case happens in two different situations: 1. when matching/try-withing on a pattern with an effect as a sub-pattern and 2. whenever an effect pattern appears in another kind of case, even if this pattern is actually a first level pattern.

This makes for a sub-optimal error messages when effects are matched against in an inappropriate place, for example:

```ocaml
# let f = function effect E, _ -> ();;
Error: Effect patterns must be at the top level of a match case.
```

A message similar as the one used for exeception would be more useful:

```ocaml
# let f = function exception _ -> ();;
Error: Exception patterns are not allowed in this position.
```

Reply:
Would you be able to propose a fix? Perhaps as a PR against https://github.com/avsm/ocaml/tree/effect-syntax?

Reply:
> Would you be able to propose a fix? Perhaps as a PR against https://github.com/avsm/ocaml/tree/effect-syntax?

I started investigating a fix, which I think will be a bit larger than I first hoped. The current implementation types effect patterns in complete isolation from other cases, and it feels wrong to to that. Additionally they are treated as `Value` pattern but they really should be treated as `Computation` patterns, citing @gasche who introduced this distinction in https://github.com/ocaml/ocaml/pull/8970:

> Value patterns match on a value (the result of computation), while
computation patterns handle the effects (hint hint) of
a computation. The only forms of computation patterns in OCaml _today_
are value patterns and exception patterns (exception p).

This lead me to think that adding effects as a `Value` case list in the `Texp_match` node is a bad choice:
```diff
-  | Texp_match of expression * computation case list * partial
+  | Texp_match of expression * computation case list * value case list * partial
```

My plan is to:
- remove the ad-hoc "value cases list" from the `Texp_match` 
- actually introduces a Typedtree node `Tpat_effect` (which should also be useful for Merlin)
- see if these changes lead to a better integration into the flow of `Typecore` (and lead to a fix of my inital concern)

Meanwhile, I am curious to know why the initial implementation chose to not introduce a Typedtree node for effect patterns ? Is there some subtlety I am missing ?

Reply:
> I am curious to know why the initial implementation chose not to introduce a Typedtree node for effect patterns ? Is there some subtlety I am missing ? 

It’s what I mentioned in an earlier comment: this PR predates #1568 and wasn’t updated during rebasing. So it still uses the old approach of having separate lists of cases.

Reply:
> It looks like I just lost my purpose 😅

Thanks @voodoos for the exploratory work. You are in a great position to review this change :-)